### PR TITLE
Support P4OVS build modes

### PR DESCRIPTION
**Pull request by ffoulkes**
### Overview

This PR modifies P4 Control Plane as part of issue https://github.com/ipdk-io/networking-recipe/issues/516.

It adds support for the newly defined P4OVS build modes.

  - In `p4ovs` (legacy) mode, OVS is built first. It creates static libraries for the `ovs-vswitchd` and `ovs-testcontroller` applications. P4CP is built second. It compiles OVSP4RT and links it with the OVS libraries to create the applications.
  - In `ovsp4rt` mode, P4CP is built first. It creates an OVSP4RT shared library. OVS is built second. It compiles `ovs-vswitchd` and `ovs-testcontroller`, and links them with the OVSP4RT library to create the applications.
 
### Changes in this PR

- Updated the cmake listfiles to support a `P4OVS_MODE` variable, which specifies the P4OVS build mode.
- Modified the P4CP cmake listfile to distinguish between legacy and non-legacy P4OVS build modes.
- Modified `make-all.sh` to accept a new `--p4ovs=MODE` parameter that specifies the P4OVS build mode. It passes the mode to the P4CP and OVS cmake builds via the `P4OVS_MODE` cmake variable.
- Modified the OVS cmake listfile to accept the `P4OVS_MODE` cmake variable and translate it to a `--with-p4ovs` or `--with-ovsp4rt` autotools configuration option.
- Modified `ovs/make-ovs.sh` to accept a the `--p4ovs=MODE` parameter and pass its value to the OVS cmake build via the `P4OVS_MODE` cmake variable. Also implemented `--help` and `--prefix` options.
- Updated the OVS submodule reference to a commit that supports the `--with-ovsp4rt` configuration option. This feature was implemented in https://github.com/ipdk-io/ovs/pull/124 and subsequent PRs.

### Compatibility

- The `P4OVS_MODE` cmake variable defaults to `p4ovs` (legacy mode).
- The `make-all.sh` and `ovs/make-ovs.sh` scripts default to `p4ovs` mode.
- The OVS cmake listfile continues to support the `P4OVS` variable.

Successful mev-ts build: https://github.com/intel-innersource/networking.ipu.mountevans.manifests/pull/9541.

### CODE REVIEW

File: make-all.sh [hunk 1]
```diff
@@ -232,18 +265,22 @@ while true ; do
...
-        _WITH_OVSP4RT=FALSE
+        _WITH_OVSP4RT=OFF
+        _with_ovsp4rt=0
         shift ;;
     --no-rpath)
         _RPATH=OFF
         shift ;;
+    --p4ovs)
+        # convert to uppercase
+        _P4OVS_MODE=${2^^}
```
Comment (5abeel) [inline make-all.sh:282]: Not familiar with what the `^^` does

File: make-all.sh [hunk 2]
```diff
@@ -80,28 +93,42 @@ print_help() {
...
+    echo "  ${_TARGET_TYPE:2}"
+    [ -n "${_TOOLCHAIN_FILE}" ] && echo "  ${_TOOLCHAIN_FILE:2}"
+    [ -n "${_WITH_KRNLMON}" ] && echo "  ${_WITH_KRNLMON:2}"
+    [ -n "${_WITH_OVSP4RT}" ] && echo "  ${_WITH_OVSP4RT:2}"
+
+    echo ""
+    echo "OvS options:"
+    [ -n "${_P4OVS_MODE}" ] && echo "  ${_P4OVS_MODE:2}"
+    [ -n "${_PKG_CONFIG_PATH}" ] && echo "  PKG_CONFIG_PATH=${_PKG_CONFIG_PATH}"
+    if [ ${_ovs_first} -ne 0 ]; then
```
Comment (5abeel) [inline make-all.sh:119]: Curious why we want this control of building OVS first or last. Is this for the ovs-p4rt dependency or somethign else?

File: ovs/CMakeLists.txt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # CMake build file for Open vSwitch
 #
-# Copyright 2022-2023 Intel Corporation
+# Copyright 2022-2024 Intel Corporation
```
Comment (5abeel) [inline ovs/CMakeLists.txt:3]: Update copyright date in other files of this PR (if they havent already been updated to 2024)

**Review by 5abeel — APPROVED** [review]
Some minor questions and comments. Rest LGTM

File: make-all.sh (hunk 1 above)
Comment (ffoulkes) [reply make-all.sh:282]: It converts a string to uppercase. `,,` converts it to lowercase. 😎

File: make-all.sh (hunk 2 above)
Comment (ffoulkes) [reply make-all.sh:119]: The immediate problem is that I need OvS to build `ovs-vswitchd` and `ovs-testcontroller` itself, against the `ovsp4rt_stubs` library, so I can run the OVS test suite against the P4 version of OVS.

The less immediate (but more important) problem is that the current build order is fundamentally wrong. networking-recipe doesn't depend on OVS; OVS depends on ovs-p4rt, which is built as part of networking-recipe.

I goofed two years ago, and am trying to put things right before upstreaming OVS.

For more information, see https://github.com/ipdk-io/networking-recipe/issues/516.

File: ovs/CMakeLists.txt (hunk 3 above)
Comment (ffoulkes) [reply ovs/CMakeLists.txt:3]: I think this is the last hold-out.

**Pull request merged**